### PR TITLE
Add stack trace for invariant violation to error message

### PR DESCRIPTION
**Pull request by gerben-stavenga**
### Description

The stacktrace proved useful in the SEV and it might be useful to have it standard in debug mode, for future incidents

### <samp>🤖 Generated by Copilot at 99062dd</samp>

This pull request improves error handling and debugging in `aptos-move`. It refactors the `session.rs` file to use a closure for errors and adds assertions and stack traces for invariant violations in `move-binary-format`.

### CODE REVIEW

File: aptos-move/aptos-vm/src/move_vm_ext/session.rs [hunk 1]
```diff
@@ -167,8 +167,10 @@ where
     ) -> VMResult<(MoveChangeSet, MoveChangeSet)> {
         // The use of this implies that we could theoretically call unwrap with no consequences,
         // but using unwrap means the code panics if someone can come up with an attack.
-        let common_error = PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
-            .finish(Location::Undefined);
+        let common_error = || {
+            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                .finish(Location::Undefined)
+        };
```
Comment (davidiw) [inline aptos-move/aptos-vm/src/move_vm_ext/session.rs:173]: so fancy

File: third_party/move/move-binary-format/src/errors.rs [hunk 2]
```diff
@@ -309,10 +309,21 @@ impl PartialVMError {
     }
 
     pub fn new(major_status: StatusCode) -> Self {
+        debug_assert!(major_status != StatusCode::EXECUTED);
```
Comment (davidiw) [inline third_party/move/move-binary-format/src/errors.rs:312]: we had typically not enabled these with the logic that we don't want different behavior from local and tests, but my memory isn't great either...

**Review by davidiw — COMMENTED** [review]
why only with debug assertions? is there a way to explicitly tell the logger to include the backtrace? @JoshLind

File: third_party/move/move-binary-format/src/errors.rs (hunk 2 above)
Comment (gerben-stavenga) [reply third_party/move/move-binary-format/src/errors.rs:312]: Typically debug asserts are used to document preconditions, but okay to confirm to a aptos style guide if this is not accepted.

**Review by zekun000 — COMMENTED** [review]
we probably need to avoid speculatively resulting this error first, otherwise it's going to be expensive

File: third_party/move/move-binary-format/src/errors.rs [hunk 3]
```diff
@@ -329,7 +340,14 @@ impl PartialVMError {
         self
     }
 
-    pub fn with_message(mut self, message: String) -> Self {
+    pub fn with_message(mut self, mut message: String) -> Self {
+        if cfg!(debug_assertions)
+            && self.0.major_status == StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR
```
Comment (runtian-zhou) [inline third_party/move/move-binary-format/src/errors.rs:345]: I was wondering if we should just append at all times.

Comment (gerben-stavenga) [reply third_party/move/move-binary-format/src/errors.rs:345]: I'm not sure if message goes into the block hash?

**Comment by gerben-stavenga** [comment]
> why only with debug assertions? is there a way to explicitly tell the logger to include the backtrace? @JoshLind



> we probably need to avoid speculatively resulting this error first, otherwise it's going to be expensive

I don't think even in speculative execution we should ever throw invariant error.

**Comment by zekun000** [comment]
> I don't think even in speculative execution we should ever throw invariant error.

Yes, we shouldn't but not sure if we already do. For example I'm changing this from unknown invariant error to storage error https://github.com/aptos-labs/aptos-core/pull/7877

File: third_party/move/move-binary-format/src/errors.rs (hunk 3 above)
Comment (vgao1996) [reply third_party/move/move-binary-format/src/errors.rs:345]: 1. We should append at all times.
2. I don't think this goes into the block hash -- only the status code does. @msmouse is this correct?

File: third_party/move/move-binary-format/src/errors.rs [hunk 4]
```diff
@@ -330,10 +330,19 @@ impl PartialVMError {
     }
 
     pub fn new(major_status: StatusCode) -> Self {
+        debug_assert!(major_status != StatusCode::EXECUTED);
+        let message = if major_status == StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR {
+            Some(format!(
+                "UNKNOWN_INVARIANT_VIOLATION_ERROR at {:?}",
+                std::backtrace::Backtrace::force_capture()
```
Comment (vgao1996) [inline third_party/move/move-binary-format/src/errors.rs:337]: Let's probably limit the number of frames we record using https://docs.rs/backtrace/latest/backtrace/

Comment (gerben-stavenga) [reply third_party/move/move-binary-format/src/errors.rs:360]: Done

**Review by runtian-zhou — APPROVED** [review]

**Pull request merged**